Race detected: grpc - Test/ServiceConfigErrorRPC

Seen on 1.17 with -race:

https://github.com/grpc/grpc-go/runs/3435757932#step:8:183

```
==================
WARNING: DATA RACE
Write at 0x00c0000e48c8 by goroutine 56:
  google.golang.org/grpc/internal/grpctest.(*tLogger).Update()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:126 +0x1ae
  google.golang.org/grpc/internal/grpctest.Tester.Setup()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/grpctest.go:50 +0x44
  google.golang.org/grpc.(*s).Setup()
      <autogenerated>:1 +0x44
  runtime.call16()
      /opt/hostedtoolcache/go/1.17.0/x64/src/runtime/asm_amd64.s:625 +0x48
  reflect.callMethod()
      <autogenerated>:1 +0x64
  reflect.methodValueCall()
      /opt/hostedtoolcache/go/1.17.0/x64/src/reflect/asm_amd64.s:80 +0x70
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x47

Previous read at 0x00c0000e48c8 by goroutine 93:
  google.golang.org/grpc/internal/grpctest.(*tLogger).log()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:99 +0x39a
  google.golang.org/grpc/internal/grpctest.(*tLogger).InfoDepth()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:195 +0x64
  google.golang.org/grpc/internal/grpclog.InfoDepth()
      /home/runner/work/grpc-go/grpc-go/internal/grpclog/grpclog.go:35 +0x85
  google.golang.org/grpc/grpclog.(*componentData).InfoDepth()
      /home/runner/work/grpc-go/grpc-go/grpclog/component.go:36 +0x174
  google.golang.org/grpc/internal/channelz.Infof()
      /home/runner/work/grpc-go/grpc-go/internal/channelz/logging.go:50 +0x19a
  google.golang.org/grpc.(*connectivityStateManager).updateState()
      /home/runner/work/grpc-go/grpc-go/clientconn.go:436 +0x1ba
  google.golang.org/grpc.(*ClientConn).Close()
      /home/runner/work/grpc-go/grpc-go/clientconn.go:1067 +0x165
  google.golang.org/grpc.s.TestResolverServiceConfigWhileClosingNotPanic·dwrap·113()
      /home/runner/work/grpc-go/grpc-go/clientconn_test.go:668 +0x39

Goroutine 56 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x726
  google.golang.org/grpc/internal/grpctest.RunSubTests()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/grpctest.go:99 +0x54a
  google.golang.org/grpc.Test()
      /home/runner/work/grpc-go/grpc-go/grpc_test.go:32 +0x3c
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x47

Goroutine 93 (finished) created at:
  google.golang.org/grpc.s.TestResolverServiceConfigWhileClosingNotPanic()
      /home/runner/work/grpc-go/grpc-go/clientconn_test.go:668 +0x573
  google.golang.org/grpc.(*s).TestResolverServiceConfigWhileClosingNotPanic()
      <autogenerated>:1 +0x44
  runtime.call16()
      /opt/hostedtoolcache/go/1.17.0/x64/src/runtime/asm_amd64.s:625 +0x48
  reflect.callMethod()
      <autogenerated>:1 +0x64
  reflect.methodValueCall()
      /opt/hostedtoolcache/go/1.17.0/x64/src/reflect/asm_amd64.s:80 +0x70
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0000e48d0 by goroutine 56:
  google.golang.org/grpc/internal/grpctest.(*tLogger).Update()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:127 +0x214
  google.golang.org/grpc/internal/grpctest.Tester.Setup()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/grpctest.go:50 +0x44
  google.golang.org/grpc.(*s).Setup()
      <autogenerated>:1 +0x44
  runtime.call16()
      /opt/hostedtoolcache/go/1.17.0/x64/src/runtime/asm_amd64.s:625 +0x48
  reflect.callMethod()
      <autogenerated>:1 +0x64
  reflect.methodValueCall()
      /opt/hostedtoolcache/go/1.17.0/x64/src/reflect/asm_amd64.s:80 +0x70
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x47

Previous read at 0x00c0000e48d0 by goroutine 93:
  runtime.racereadrange()
      <autogenerated>:1 +0x1b
  google.golang.org/grpc/internal/grpctest.(*tLogger).InfoDepth()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:195 +0x64
  google.golang.org/grpc/internal/grpclog.InfoDepth()
      /home/runner/work/grpc-go/grpc-go/internal/grpclog/grpclog.go:35 +0x85
  google.golang.org/grpc/grpclog.(*componentData).InfoDepth()
      /home/runner/work/grpc-go/grpc-go/grpclog/component.go:36 +0x174
  google.golang.org/grpc/internal/channelz.Infof()
      /home/runner/work/grpc-go/grpc-go/internal/channelz/logging.go:50 +0x19a
  google.golang.org/grpc.(*connectivityStateManager).updateState()
      /home/runner/work/grpc-go/grpc-go/clientconn.go:436 +0x1ba
  google.golang.org/grpc.(*ClientConn).Close()
      /home/runner/work/grpc-go/grpc-go/clientconn.go:1067 +0x165
  google.golang.org/grpc.s.TestResolverServiceConfigWhileClosingNotPanic·dwrap·113()
      /home/runner/work/grpc-go/grpc-go/clientconn_test.go:668 +0x39

Goroutine 56 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x726
  google.golang.org/grpc/internal/grpctest.RunSubTests()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/grpctest.go:99 +0x54a
  google.golang.org/grpc.Test()
      /home/runner/work/grpc-go/grpc-go/grpc_test.go:32 +0x3c
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x47

Goroutine 93 (finished) created at:
  google.golang.org/grpc.s.TestResolverServiceConfigWhileClosingNotPanic()
      /home/runner/work/grpc-go/grpc-go/clientconn_test.go:668 +0x573
  google.golang.org/grpc.(*s).TestResolverServiceConfigWhileClosingNotPanic()
      <autogenerated>:1 +0x44
  runtime.call16()
      /opt/hostedtoolcache/go/1.17.0/x64/src/runtime/asm_amd64.s:625 +0x48
  reflect.callMethod()
      <autogenerated>:1 +0x64
  reflect.methodValueCall()
      /opt/hostedtoolcache/go/1.17.0/x64/src/reflect/asm_amd64.s:80 +0x70
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.0/x64/src/testing/testing.go:1306 +0x47
==================
--- FAIL: Test (20.01s)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+3.854927ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+3.994528ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+4.112329ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+4.22613ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+4.318531ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+4.422031ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+4.544232ms)
    tlogger.go:99: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[] 0xc0000630e0 <nil>}  (t=+4.660633ms)
    tlogger.go:99: clientconn.go:728 [core] ClientConn switching balancer to "round_robin"  (t=+4.719633ms)
    tlogger.go:99: clientconn.go:748 [core] Channel switches to new LB policy "round_robin"  (t=+4.768834ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to TRANSIENT_FAILURE  (t=+4.874034ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+4.949235ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+5.049636ms)
    tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+5.123036ms)
    --- FAIL: Test/ServiceConfigErrorRPC (0.00s)
        tlogger.go:99: clientconn.go:252 [core] parsed scheme: "whatever"  (t=+150.401µs)
        tlogger.go:99: service_config.go:230 [core] grpc: parseServiceConfig error unmarshaling bad config due to invalid character 'b' looking for beginning of value  (t=+229.101µs)
        tlogger.go:99: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[] 0xc0000638e0 <nil>}  (t=+286.002µs)
        tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to TRANSIENT_FAILURE  (t=+366.002µs)
        tlogger.go:99: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+469.603µs)
        testing.go:1152: race detected during execution of test
    testing.go:1152: race detected during execution of test
FAIL
FAIL	google.golang.org/grpc	38.961s
```

--- a/internal/grpctest/tlogger.go
+++ b/internal/grpctest/tlogger.go
@@ -49,11 +49,11 @@ const (
 
 type tLogger struct {
 	v           int
-	t           *testing.T
-	start       time.Time
 	initialized bool
 
-	m      sync.Mutex // protects errors
+	mu     sync.Mutex // guards t, start, and errors
+	t      *testing.T
+	start  time.Time
 	errors map[*regexp.Regexp]int
 }
 
@@ -76,6 +76,8 @@ func getCallingPrefix(depth int) (string, error) {
 
 // log logs the message with the specified parameters to the tLogger.
 func (g *tLogger) log(ltype logType, depth int, format string, args ...interface{}) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	prefix, err := getCallingPrefix(callingFrame + depth)
 	if err != nil {
 		g.t.Error(err)
@@ -119,14 +121,14 @@ func (g *tLogger) log(ltype logType, depth int, format string, args ...interface
 // Update updates the testing.T that the testing logger logs to. Should be done
 // before every test. It also initializes the tLogger if it has not already.
 func (g *tLogger) Update(t *testing.T) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	if !g.initialized {
 		grpclog.SetLoggerV2(TLogger)
 		g.initialized = true
 	}
 	g.t = t
 	g.start = time.Now()
-	g.m.Lock()
-	defer g.m.Unlock()
 	g.errors = map[*regexp.Regexp]int{}
 }
 
@@ -141,20 +143,20 @@ func (g *tLogger) ExpectError(expr string) {
 
 // ExpectErrorN declares an error to be expected n times.
 func (g *tLogger) ExpectErrorN(expr string, n int) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	re, err := regexp.Compile(expr)
 	if err != nil {
 		g.t.Error(err)
 		return
 	}
-	g.m.Lock()
-	defer g.m.Unlock()
 	g.errors[re] += n
 }
 
 // EndTest checks if expected errors were not encountered.
 func (g *tLogger) EndTest(t *testing.T) {
-	g.m.Lock()
-	defer g.m.Unlock()
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	for re, count := range g.errors {
 		if count > 0 {
 			t.Errorf("Expected error '%v' not encountered", re.String())
@@ -165,8 +167,6 @@ func (g *tLogger) EndTest(t *testing.T) {
 
 // expected determines if the error string is protected or not.
 func (g *tLogger) expected(s string) bool {
-	g.m.Lock()
-	defer g.m.Unlock()
 	for re, count := range g.errors {
 		if re.FindStringIndex(s) != nil {
 			g.errors[re]--

